src/dune_engine/build_system.ml

```diff
@@ -957,36 +957,46 @@ end = struct
     in
     (* Compute the set of targets and the set of source files that must not be
        copied *)
-    let source_files_to_ignore =
-      List.fold_left rules ~init:Path.Build.Set.empty
-        ~f:(fun acc_ignored { Rule.targets; mode; loc; _ } ->
+    let source_files_to_ignore, source_dirs_to_ignore =
+      List.fold_left rules ~init:(Path.Build.Set.empty, Path.Build.Set.empty)
+        ~f:(fun (acc_files, acc_dirs) { Rule.targets; mode; loc; _ } ->
           (* CR-someday amokhov: Remove this limitation. *)
           let directory_targets_not_supported ~targets =
             if not (Path.Build.Set.is_empty targets.Targets.Validated.dirs) then
               User_error.raise ~loc
                 [ Pp.text "Directory targets are not supported for this mode" ]
           in
           match mode with
-          | Promote { only = None; _ }
           | Ignore_source_files ->
             directory_targets_not_supported ~targets;
-            Path.Build.Set.union targets.files acc_ignored
-          | Promote { only = Some pred; _ } ->
-            directory_targets_not_supported ~targets;
-            let to_ignore =
-              Path.Build.Set.filter targets.files ~f:(fun target ->
+            (Path.Build.Set.union targets.files acc_files, acc_dirs)
+          | Promote { only; _ } ->
+            let files, dirs =
+              match only with
+              | None -> (targets.files, targets.dirs)
+              | Some pred ->
+                let is_promoted target =
                   Predicate_lang.Glob.exec pred
                     (Path.reach (Path.build target) ~from:(Path.build dir))
-                    ~standard:Predicate_lang.any)
+                    ~standard:Predicate_lang.any
+                in
+                ( Path.Build.Set.filter targets.files ~f:is_promoted
+                , Path.Build.Set.filter targets.dirs ~f:is_promoted )
             in
-            Path.Build.Set.union to_ignore acc_ignored
+            ( Path.Build.Set.union files acc_files
+            , Path.Build.Set.union dirs acc_dirs )
           | Standard
           | Fallback ->
-            acc_ignored)
+            (acc_files, acc_dirs))
     in
     let source_files_to_ignore =
-      Path.Build.Set.to_list source_files_to_ignore
-      |> Path.Source.Set.of_list_map ~f:Path.Build.drop_build_context_exn
+      Path.Build.Set.to_list_map ~f:Path.Build.drop_build_context_exn
+        source_files_to_ignore
+      |> Path.Source.Set.of_list
+    in
+    let source_dirs_to_ignore =
+      Path.Build.Set.to_list_map ~f:Path.Build.basename source_dirs_to_ignore
+      |> String.Set.of_list
```

Comment:
Probably not worth changing, but it is bit niggling that `source_files_to_ignore` is a `Path.Source.Set.t` while `source_dirs_to_ignore` is a `String.Set.t`. Feels like it should be symetrical.

Reply:
Agreed, I'll change that.

Reply:
Actually, the reason file and directory sets are treated differently is that they are passed to `create_copy_rules` and `Subdir_set.These`, respectively, and they have different APIs. So, I'm keeping this unchanged to avoid growing this PR too much.